main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1114,6 +1116,50 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
       }
   }
 
+  /**
+    * Performs name resolution on the given list of derivations `derives0`.
+    */
+  def resolveDerivations(derives0: List[Name.QName], ns0: Name.NName, root: NamedAst.Root): Validation[List[Symbol.ClassSym], ResolutionError] = {
+    val derivesVal = Validation.traverse(derives0)(resolveDerivation(_, ns0, root))
+    flatMapN(derivesVal) {
+      derives =>
+        val mapping = derives0.zip(derives)
+        val duplicatePair = mapping.combinations(2).collectFirst {
+          case List((qname1, sym1), (qname2, sym2)) if (sym1 == sym2) => (sym1, qname1.loc, qname2.loc)
+        }
+        duplicatePair match {
+          case Some((sym, loc1, loc2)) => ResolutionError.DuplicateDerivation(sym, loc1, loc2).toFailure
+          case None => derives.toSuccess
+        }
```

Comment:
I was upset with the solution, so I spent way too long trying to come up with something better.

Here is my attempt (sketch):

```scala
  def foo(): Unit = {
    val derives: List[(String, String)] = ("theName", "theLoc") :: Nil

    val clashes = for {
      (qname1, loc1) <- derives
      (qname2, loc2) <- derives;
      if qname1 == qname2
    } yield (qname1, loc1, loc2)
    
    sequenceM(Success(derives) :: clashes.map(ResolutionError.DuplicateDerivation))
  }
```

Reply:
My initial attempt was something like that, based on the check for duplicate type constraints (https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala#L206)

A few issues: 
1. We need to compare `sym`s rather than `qname`s because `qname` equality is not well-defined.
2. Sym location is at declaration rather than use, so we need to use the `qname` location.
3. We need to be sure not to compare a sym against itself.
4. If we get all pairs then we end up with a polynomial explosion of errors, which seemed bad to me. (But not as bad as the exponential that I was originally fearing)

`list.combinations` takes care of 3 by itself, and `collectFirst` takes care of 4.

That all said, the double-for is certainly possible, though I don't think it will end up being as pretty as what you propose :)